scripts: enhance the release note extraction script

knz opened the pull request:
.. and add unit tests.

Fixes #24600.
Fixes #24979.

The following have been improved:

- if a PR contains multiple commits, and only some of them contain
  a release note, then the remaining commits will not be listed
  any more in the section "changes without release note".
  This aligns the behavior with what is allowed in the contributor guide.
- the script now recognizes the new Github marker "Co-authored-by" in
  commit messages.
- if a single commit contains multiple release notes, they will now be
  correctly separated into multiple entries.
- the script now finds/recognizes commits that have been committed
  to the main branch without a fork/merge.
- intermediate pulls/merges into feature branches are now properly
  handled.
- newline characters at the end of a PR title are now properly
  stripped.
- the release notes are now listed in the order they were committed,
  instead of reverse order. This makes the list more readable (the reader
  can "follow along" with the changes).

Release note: None

cc @jseldess

knz committed:
scripts: enhance the release note extraction script

.. and add unit tests.

The following have been improved:

- the first word of each release note is capitalized during
  the collection.
- if a PR contains multiple commits, and only some of them contain
  a release note, then the remaining commits will not be listed
  any more in the section "changes without release note".
  This aligns the behavior with what is allowed in the contributor guide.
- the script now recognizes the new Github marker "Co-authored-by" in
  commit messages.
- if a single commit contains multiple release notes, they will now be
  correctly separated into multiple entries.
- the script now finds/recognizes commits that have been committed
  to the main branch without a fork/merge.
- intermediate pulls/merges into feature branches are now properly
  handled.
- newline characters at the end of a PR title are now properly
  stripped.
- the release notes are now listed in the order they were committed,
  instead of reverse order. This makes the list more readable (the reader
  can "follow along" with the changes).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24978)
<!-- Reviewable:end -->

knz commented:
I'm still working on this (to clean up the python code, add more tests) but I wanted to get it out for some early feedback.

Nikhil I'll take suggestions as to how you recommend to get this tested on CI when the script is modified (the tests don't need to run otherwise).

knz commented:
@jseldess when I have advanced a bit on this I'll suggest we try it out on some of the past releases to check that no entries have been forgotten (I hope none have, but I suspect some were)

jseldess commented:
Thanks for these improvements, @knz! Let me know when you want me to do testing.

knz commented:
This is now ready for testing & review!

couchand commented on scripts/release-notes.py:
```diff
@@ -169,6 +169,9 @@
 form3 = r':(?P<cat3>)'
 relnote = re.compile(r'(?:^|[\n\r])[rR]elease [nN]otes? *(?:' + form1 + '|' + form2 + '|' + form3 + r') *(?P<note>.*)$', flags=re.S)
 
+coauthor = re.compile(r'^Co-authored-by: (?P<name>[^<].*) <(?P<email>.*)>', flags=re.M)
```
This regex looks suspicious to me, specifically the `(?P<name>[^<].*)` group

couchand commented on scripts/release-notes.py:
```diff
@@ -169,6 +169,9 @@
 relnote = re.compile(r'(?:^|[\n\r])[rR]elease [nN]otes? *(?:' + form1 + '|' + form2 + '|' + form3 + r') *(?P<note>.*)$', flags=re.S)
 
+coauthor = re.compile(r'^Co-authored-by: (?P<name>[^<].*) <(?P<email>.*)>', flags=re.M)
+fixannot = re.compile(r'^(Fix(es|ed)?|Close(d|s)?) #', flags=re.M)
```
maybe s/F/[Ff]/ and s/C/[Cc]/?

couchand commented on scripts/release-notes.py:
```diff
@@ -269,40 +275,111 @@
+    for line in msglines:
+        m = coauthor.search(line)
+        if m is not None:
+            # We ignore co-authored-by lines.
```
it seems like coauthor and fix/close lines should end the current release note?

couchand commented on scripts/release-notes.py:
```diff
@@ -269,40 +275,111 @@
+        if firstline.lower() == 'none':
+            # Release note: none - there's no note yet.
+            continue
+        foundnote = True
```
should this line be above the previous block?  `Release note: None` is still a note, no?

couchand commented on scripts/release-notes.py:
```diff
@@ -395,68 +472,94 @@ def analyze_pr(merge, pr):
+def analyze_standalone_commit(commit):
+    # Some random out-of-branch commit. Let's not forget them.
+    authors = collect_authors(commit)
+    title = commit.message.split('\n')[0].strip()
```
`split('\n', 1)`?

couchand commented on scripts/release-notes.py:
```diff
@@ -395,68 +472,94 @@ def analyze_pr(merge, pr):
+    # Some random out-of-branch commit. Let's not forget them.
+    authors = collect_authors(commit)
+    title = commit.message.split('\n')[0].strip()
+    item = {'authors': ', '.join(sorted(authors)),
```
ok i've seen this little snippet enough times that i'll suggest it get extracted

couchand commented on scripts/release-notes.py:
```diff
@@ -395,68 +472,94 @@ def analyze_pr(merge, pr):
+            'pr': '#unknown',
+            'prtitle': title,
+            'note': None}
+    missing_release_notes.append(item)
```
couldn't the commit message have a release note?

couchand commented on scripts/release-notes/test1.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+set -eux
+
+. common.sh
```
is this PR missing `common.sh`?

couchand reviewed (approved):
Thanks for developing that testing stuff!  I think `common.sh` is missing, otherwise LGTM.  I left a variety of random comments, feel free to ignore them at your leisure.

knz commented:
Thanks for the review. Sorry I had forgotten common.sh. It's included now. I'll process your other comments tomorrow.

knz committed:
githook: add a release note linter

... to warn / recommend against release notes that are too short and
just say "fixes #nnn" without explaining what's going on.

Release note: None

couchand commented on scripts/release-notes/common.sh:
```diff
@@ -0,0 +1,105 @@
+# Initialize the repository. Tag the initial commit.
+function init_repo() {
+    git init
+    touch foo; git add foo; git commit "${flags[@]}" -m "initial"; git tag initial
```
if you wish you could `--allow-empty` here and in `make_change`

jseldess commented:
@knz, I'm not able to run the before part of the before/after comparison:

```
~/go/src/github.com/cockroachdb/cockroach$ python3 scripts/release-notes.py --from=v2.1-alpha.0000000 --until=v2.1.0-alpha.20180416 --hide-unambiguous-shas --hide-per-contributor-section > untilv2.1.0-alpha.md
Unable to find the first commit of the range.
No ref named v2.1-alpha.0000000.
```

Let me know if you have ideas.

knz commented:
@jseldess 

> I'm not able to run the before part of the before/after comparison:

I was missing a zero. Try `--from=v2.1-alpha.00000000`.

knz committed:
scripts: add rich to the release notes script

Release note: none

knz commented:
Review status: 0 of 21 files reviewed at latest revision, 9 unresolved discussions.

---

*[scripts/release-notes.py, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-172:-LAyMmMa7t7LeGiqTMkf:bi9ipn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L172)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This regex looks suspicious to me, specifically the `(?P<name>[^<].*)` group
</blockquote></details>

Nicely spotted. Fixed.

---

*[scripts/release-notes.py, line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-173:-LAyMo4NCHmOFI7fJnrW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L173)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

maybe s/F/[Ff]/ and s/C/[Cc]/?
</blockquote></details>

Done.

---

*[scripts/release-notes.py, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-309:-LAyMogc7Lny0F_Za9L2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L309)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

it seems like coauthor and fix/close lines should end the current release note?
</blockquote></details>

Done.

---

*[scripts/release-notes.py, line 337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-337:-LAyMpZm9xVaL6gNYjZ4:b-ws65pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L337)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

should this line be above the previous block?  `Release note: None` is still a note, no?
</blockquote></details>

I don't understand, can you give more words?

---

*[scripts/release-notes.py, line 554 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-554:-LAyMz2L-vR67RFo-qw3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L554)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

`split('\n', 1)`?
</blockquote></details>

Done.

---

*[scripts/release-notes.py, line 555 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-555:-LAyMzZ33VE0FX1bxh4g:b-m9xpwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L555)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

ok i've seen this little snippet enough times that i'll suggest it get extracted
</blockquote></details>

Which one?

---

*[scripts/release-notes.py, line 560 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-560:-LAyN-lm3MneqVG355x3:b-fjmlyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L560)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

couldn't the commit message have a release note?
</blockquote></details>

If they are not associated to a PR I want the docs maintainer to make an extra step and read the change to figure out where it belongs (which PR / issue). Hence this choice.

---

*[scripts/release-notes/common.sh, line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAtW05z_mD9pTI2w3PF-r2-25:-LAyN9L7-gv87eqiKXOi:b-qwa53z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7672492d82210c9cd330d546fdb159564799105/scripts/release-notes/common.sh#L25)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

if you wish you could `--allow-empty` here and in `make_change`
</blockquote></details>

TIL. What does this do?

---

*[scripts/release-notes/test1.sh, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlArn-r1-4:-LAyNBEF4jfoFNgNlYZV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes/test1.sh#L4)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

is this PR missing `common.sh`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24978)*
<!-- Sent from Reviewable.io -->

couchand commented:
Reviewed 11 of 19 files at r1, 1 of 1 files at r2, 7 of 8 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[scripts/release-notes.py, line 337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-337:-LB1bdkdD5rIq-kIqRQR:bsdiq54) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L337)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't understand, can you give more words?
</blockquote></details>

I mean, if you see `Release note: None`, that should still set `foundnote = True`.  If a PR has only commits with `Release note: None`, it's not missing release notes, right?

---

*[scripts/release-notes.py, line 555 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-555:-LB1c0xlB3XfmafzTJ9h:brt3czw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L555)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Which one?
</blockquote></details>

the `', '.join(sorted(authors))` part.  three different places are doing that, and the results are later compared to uniqify.  Perhaps this data structure should just keep the whole list of authors, and then at a later point they get stringified all at once, or alternatively, just have all three places call a method to stringify.

---

*[scripts/release-notes.py, line 560 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAjWnQZ2waz-CZYlAri-r1-560:-LB1cS7_979BOrMtws2u:b-abrsvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3a0ed192b7001b17a81df8badb739363a38f9ae/scripts/release-notes.py#L560)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

If they are not associated to a PR I want the docs maintainer to make an extra step and read the change to figure out where it belongs (which PR / issue). Hence this choice.
</blockquote></details>

Fair point.  It could still extract the release notes to make their job a little easier, and put them in a different section maybe, to make everything clearer?

---

*[scripts/release-notes/common.sh, line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LAtW05z_mD9pTI2w3PF-r2-25:-LB1cwsf20TN4FJ6bzSS:b1a6cly) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7672492d82210c9cd330d546fdb159564799105/scripts/release-notes/common.sh#L25)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

TIL. What does this do?
</blockquote></details>

Just lets the commit go through with no changes in it, so you don't have to bother with all the `touch`ing.  Usually disallowed because it doesn't really do anything useful.  Sometimes I make the first commit on a repo with `--allow empty` because it makes some history rewriting easier later on, but otherwise I've never seen a use for it.  Seems like a good fit here, though.

---

*[scripts/release-notes/common.sh, line 102 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-LB1dwV86lHwt1k61dU2:-LB1dwVA6XsnN92uOlb6:b7tgjli) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcc19670647aaa958b4f34ae39e6ee41d48390/scripts/release-notes/common.sh#L102)):*
> ```Shell
> 	diff -u $t.notes.txt $t.notes.ref.txt
>     else
> 	mv -f $t.notes.txt $t.notes.ref.txt
> ```

looks like some whitespace ugliness on each `diff` and `mv` line here...

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24978)*
<!-- Sent from Reviewable.io -->

couchand commented:
Ok, I think the rest of my comments are minor enough.  <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24978#-:-LB1gO_p6ggDfuV5rSCo:b-t86k21)*
<!-- Sent from Reviewable.io -->

knz commented:
@jseldess let me know when/if this works for you.

jseldess commented:
@knz, seeing some errors:

```
~/go/src/github.com/cockroachdb/cockroach$ python3 scripts/release-notes.py --from=v2.1-alpha.00000000 --until v2.1.0-alpha.20180416 --hide-unambiguous-shas --hide-per-contributor-section > from_v2.1-alpha.00000000_knz.md
Changes from
0891a879e64d6bc3c4e16ff525a98982f0056d9b ("Merge pull request #22776 from petermattis/pmattis/roachtest-exit-code", Fri Feb 16 14:09:35 2018)
until
f54870592a75b62bf8b04ab6eb1c74442b57164f ("Merge #24594", Mon Apr  9 22:04:56 2018)
................./

Traceback (most recent call last):
  File "scripts/release-notes.py", line 709, in <module>
    print("- [%(pr)s][%(pr)s] [%(sha)s][%(sha)s] %(prtitle)s" % item, "(%s)" % authors)
KeyError: 'prtitle'
~/go/src/github.com/cockroachdb/cockroach$ python3 scripts/release-notes.py --from=v2.1-alpha.00000000 --until v2.1.0-alpha.20180416 --hide-unambiguous-shas --hide-per-contributor-section > from_v2.1-alpha.00000000_knz.md
Changes from
0891a879e64d6bc3c4e16ff525a98982f0056d9b ("Merge pull request #22776 from petermattis/pmattis/roachtest-exit-code", Fri Feb 16 14:09:35 2018)
until
f54870592a75b62bf8b04ab6eb1c74442b57164f ("Merge #24594", Mon Apr  9 22:04:56 2018)
................./

Traceback (most recent call last):
  File "scripts/release-notes.py", line 709, in <module>
    print("- [%(pr)s][%(pr)s] [%(sha)s][%(sha)s] %(prtitle)s" % item, "(%s)" % authors)
KeyError: 'prtitle'
```

jseldess commented:
LGTM!

knz commented:
Thanks! I've filed #25180 for the remaining conversation.

bors r+
